crates/volta-core/src/tool/package/fetch.rs

```diff
@@ -79,9 +76,90 @@ fn load_cached_distro(file: &Path, shasum_file: &Path) -> Option<Box<dyn Archive
-    Tarball::fetch(url, path).with_context(download_tool_error(spec, url.to_string()))
+fn fetch_remote_distro(
+    path: &Path,
+    name: &str,
+    details: &PackageDetails,
+    session: &mut Session,
+) -> Fallible<Box<Archive>> {
+    ensure_containing_dir_exists(&path).with_context(|_| ErrorDetails::ContainingDirError {
+        path: path.to_path_buf(),
+    })?;
+
+    let dir = path.parent().unwrap();
+
+    let command = npm_pack_command_for(name, &details.version.to_string()[..], session, Some(dir))?;
+    debug!("Running command: `{:?}`", command);
+
+    let spinner = progress_spinner(&format!(
+        "Downloading via {} npm pack to {}",
+        tool_version(name, details.version.to_string()),
+        dir.to_str().unwrap()
+    ));
+    let output = command.output()?;
+    spinner.finish_and_clear();
+
+    if !output.status.success() {
+        debug!(
+            "Command failed, stderr is:\n{}",
+            String::from_utf8_lossy(&output.stderr).to_string()
+        );
+        debug!("Exit code is {:?}", output.status.code());
+        throw!(ErrorDetails::NpmPackMetadataFetchError {
+            package: name.to_string(),
+        });
+    }
+
+    let response_json = String::from_utf8_lossy(&output.stdout);
+
+    let metadatas: Vec<super::serial::NpmPackData> = serde_json::de::from_str(&response_json)
+        .with_context(|_| ErrorDetails::NpmPackMetadataParseError {
+            package: name.to_string(),
+        })?;
+
+    debug!("Finding the first/only record from the output");
+    let metadata: super::serial::NpmPackData = match metadatas.iter().next() {
+        Some(data) => data.clone(),
+        // TODO: Make this be a correct error
+        None => throw!(ErrorDetails::PackageNotFound {
+            package: name.to_string()
+        }),
+    };
+
+    let tarball_from_npm_pack = dir.join(metadata.filename);
+
+    debug!("Moving the tarball to the expected path");
+    rename(tarball_from_npm_pack, path).with_context(|_| ErrorDetails::SetupToolImageError {
+        tool: name.into(),
+        version: details.version.to_string(),
+        dir: dir.to_path_buf(),
+    })?;
+
+    debug!("Attempting to load the now cached tarball from disk");
+    // TODO: Make this be a correct error
+    let distro = File::open(path).with_context(|_| ErrorDetails::PackageNotFound {
+        package: name.to_string(),
+    })?;
+
+    // TODO: Make this be a correct error
+    Tarball::load(distro).with_context(|_| ErrorDetails::PackageNotFound {
+        package: name.to_string(),
+    })
+}
+
+// build a command to run `npm pack` with json output
+fn npm_pack_command_for(
+    name: &str,
+    version: &str,
+    session: &mut Session,
+    current_dir: Option<&Path>,
+) -> Fallible<ToolCommand> {
+    let args = vec![
+        OsString::from("pack"),
+        OsString::from("--json"),
+        OsString::from(format!("{}@{}", name, version)),
+    ];
+    run::npm::command(args, session, current_dir)
```

Comment:
Instead of threading the `current_dir` through all of the unrelated `ToolCommand` code, what we can instead do is add a `current_dir` method to the `impl ToolCommand` block in `volta-core/src/run/mod.rs`:

```rust
pub(crate) fn current_dir<P: AsRef<Path>>(&mut self, dir: P) -> &mut ToolCommand {
    self.command.current_dir(dir);
    self
}
```

Since we aren't threading an `Option` through all of the sub-commands, we can change this function to take the `&Path` directly, instead of an option (since it's always used), and then here we can do:

```rust
let mut command = run::npm::command(args, session);
command.current_dir(current_dir);
command
```

Reply:
Yeah, this is what I was trying to do, and did implement once, but I had issues with command being moved and everything broke. I'm sure I just had my `&`s or my `mut`s wrong in some way, so I'll give your provided code a try!